File: Java/other/PalindromeLinkedList.java
// This program is to check if a singly linked list is a palindrome or not.
// And, A singly linked list is a linked list that can be traversed in only one direction from head to the last node, i.e Unidirectional essentially.
// A string is said to be palindrome if it reads the same backward as forward. 
// eg. racecar, madam etc.

import java.util.*;

class Node
{
  int data;
  Node next;
  public Node (int data)
  {
    this.data = data;
    this.next = null;
  }
}

class Palindrome
{
  int n;
  //Represent the head and tail of the singly linked list  
  public Node head = null;
  public Node tail = null;

  //add() will add a new node to the list  
  public void add (int data)
  {
    //Create a new node  
    Node newNode = new Node (data);

    //Then, Check if the list is empty  
    if (head == null)
      {
	//If the list turns out to be empty, both head and tail will point to new node.  
	head = newNode;
	tail = newNode;
      }
    else
      {
	//newNode will be added after tail such that tail's next will point to newNode now.  
	tail.next = newNode;
	//newNode will become new tail of the list.  
	tail = newNode;
      }
    // n will count the number of nodes present in the list. 
    n++;
  }

  //reverseList() will reverse the singly linked list and return the head of the list  
  public Node reverseList (Node temp)
  {
    Node current = temp;
    Node prev = null, next = null;

    //Swapping the previous and next nodes of each node to reverse the direction of the list.  
    while (current != null)
      {
	next = current.next;
	current.next = prev;
	prev = current;
	current = next;
      }
    return prev;
  }

  //isPalindrome() will determine whether given list is palindrome or not.  
  public void isPalindrome ()
  {
    Node current = head;
    boolean flag = true;

    //Store the mid position of the list  
    int mid = (n % 2 == 0) ? (n / 2) : ((n + 1) / 2);

    //Find the middle node in the given singly linked list  
    for (int i = 1; i < mid; i++)
      {
	current = current.next;
      }

    //Reverse the list after middle node to end. 
    Node rev = reverseList (current.next);

    //Compare nodes of first half and second half of list  
    while (head != null && rev != null)
      {
	if (head.data != rev.data)
	  {
	    flag = false;
	    break;
	  }
	head = head.next;
	rev = rev.next;
      }

    if (flag)
      System.out.println ("Given linked list is a palindrome.");
    else
      System.out.println ("Given linked list is NOT a palindrome.");
  }

  //display() will display all the nodes present in the list. 
  public void display ()
  {
    //Node current will point to head.  
    Node current = head;

    if (head == null)
      {
	System.out.println ("List is empty.");
	return;
      }
    System.out.println ("Nodes of the linked list: ");
    while (current != null)
      {
	//Prints each node by incrementing pointer.  
	System.out.print ((char)current.data + " ");
	current = current.next;
      }
    System.out.println ();
  }
}

public class Main
{
  public static void main (String[]args)
  {

    Scanner sc = new Scanner (System.in);
    Palindrome list = new Palindrome();

    //Add nodes to the list.  
    System.out.println("Enter the string: ");
    String str=sc.nextLine().trim();

    for(int i=0;i<str.length();i++)
        list.add(str.charAt(i));

    list.display ();

    //Checks whether given list is palindrome or not.  
    list.isPalindrome ();
  }
}

// Time Complexity = O(n) //This represents the number of operations to be performed by an algorithm to complete its task.
// Space Complexity = O(1) //This represents the total space taken by the algorithm with respect to the input size.
/*
Output:
Enter the string:                                                                                                         
racecar                                                                                                                   
Nodes of the linked list:                                                                                                 
r a c e c a r                                                                                                             
Given linked list is a palindrome.                                                                                        
                                     
Enter the string:                                                                                                       
candy                                                                                                                   
Nodes of the linked list:                                                                                               
c a n d y                                                                                                               
Given linked list is NOT a palindrome. 

Enter the string:                                                                                                       
Anime                                                                                                                   
Nodes of the linked list:                                                                                               
A n i m e                                                                                                               
Given linked list is NOT a palindrome.   

Enter the string:                                                                                                       
malayalam                                                                                                               
Nodes of the linked list:                                                                                               
m a l a y a l a m                                                                                                       
Given linked list is a palindrome. 
*/

File: C-Plus-Plus/other/palindrome_linked_list.cpp
//A program to check if a given singly linked list is a palindrome or not in linear time and no extra space
//The idea is to reverse the second half of the linked list, and then compare the two halfs for palindrome checking

#include <bits/stdc++.h>
using namespace std;

class ListNode {
public:
    int val;
    ListNode* next;
};

//return if the linked list is a palindrome or not
bool is_palindrome(ListNode* head)
{

    ListNode* A = head;
    int cnt = 0;

    //Count the number of nodes in the linked list
    while (A != NULL) {
        cnt++;
        A = A->next;
    }

    //If there is only one node, then its clearly a palindrome
    if (cnt == 1) {
        return 1;
    }

    //Traverse upto the node(half_end) after which we need to reverse the linked list
    int half = (cnt + 1) / 2;
    half--;

    ListNode* half_end = head;
    while (half > 0) {
        half_end = half_end->next;
        half--;
    }

    //Take 2 pointers to reverse the linked list after node half_end
    ListNode* p1 = half_end;
    ListNode* p2 = half_end->next;

    ListNode* temp = p2->next;
    p2->next = NULL;
    p1 = p2;
    p2 = temp;

    while (p2) {
        temp = p2->next;
        p2->next = p1;
        p1 = p2;
        p2 = temp;
    }

    //Join the head of the reversed linked list to the current endthat is, half_end
    half_end->next = p1;

    //Take 2 pointers and compare the 2 parts
    p2 = head;
    p1 = half_end->next;

    while (p1) {
        //If any one value is different, return 0
        if ((p2->val) != (p1->val)) {
            return 0;
        }
        else {
            p2 = p2->next;
            p1 = p1->next;
        }
    }

    return 1;
}

int main()
{
    int nodes;
    cin >> nodes;

    int node_val;
    cin >> node_val;

    ListNode* head = new ListNode();
    head->val = node_val;
    head->next = NULL;

    ListNode* temp = head;

    for (int i = 1; i < nodes; i++) {
        cin >> node_val;
        ListNode* new_node = new ListNode();
        new_node->val = node_val;
        new_node->next = NULL;
        temp->next = new_node;
    }

    if (is_palindrome(head)) {
        cout << "Yes, it is a palindrome" << endl;
    }
    else {
        cout << "No, it is not a palindrome" << endl;
    }
}

/*Test Case 1:
  n= 6
  input= 2->4->6->6->4->2
  output= Yes, it is a palindrome

  Test Case 2:
  n= 5
  input= 1->2->3->5->9
  output= No, it is not a palindrome

  Time Complexity: O(n)
  (Since, we traverse the linked list only thrice, once for counting the nodes,
   then to reverse the second half and then for checking if it is a palindrome)
  Space Complexity: O(1) (No extra space is used)*/
